tests/integration/test_dynamodb.py

```diff
@@ -28,3 +30,56 @@ def tes_can_list_tables(self):
         # Smoke test to verify we can list() the tables
         # collection.
         self.assertIsInstance(list(self.ddb.tables.all()), list)
+
+
+class TestDynamoDBTypes(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.session = boto3.session.Session(region_name='us-west-2')
+        cls.dynamodb = cls.session.resource('dynamodb')
+        cls.table_name = unique_id('boto3db')
+
+        cls.item_data = {
+            'MyHashKey': 'mykey',
+            'MyNull': None,
+            'MyBool': True,
+            'MyString': 'mystring',
+            'MyNumber': Decimal('1.25'),
+            'MyBinary': Binary(b'\x01'),
+            'MyStringSet': set(['foo']),
+            'MyNumberSet': set([Decimal('1.25')]),
+            'MyBinarySet': set([Binary(b'\x01')]),
+            'MyList': ['foo'],
+            'MyMap': {'foo': 'bar'}
+        }
+
+        cls.table = cls.dynamodb.create_table(
+            TableName=cls.table_name,
+            ProvisionedThroughput={"ReadCapacityUnits": 5,
+                                   "WriteCapacityUnits": 5},
+            KeySchema=[{"AttributeName": "MyHashKey", "KeyType": "HASH"}],
+            AttributeDefinitions=[{"AttributeName": "MyHashKey",
+                                   "AttributeType": "S"}])
+
+        waiter = cls.dynamodb.meta.client.get_waiter('table_exists')
+        waiter.wait(TableName=cls.table_name)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.table.delete()
+
+    def test_item_resource(self):
+        self.table.put_item(Item=copy.deepcopy(self.item_data))
```

Comment:
It is sort of strange that the dictionary inputted by the user will get modified (that is why I needed to send it a copy of it). I am not sure if there is a great work around this because all of these transformations happens before serialization and there is not really a way to return a new dictionary of params for the ``before-parameter-build`` botocore event.